### ISSUE TITLE
RuntimeBinderException when dynamically casting private enum type

### ISSUE THREAD

**Issue by mikernet** [dotnet/coreclr#27953]
It took me way too long to figure out what the heck was going on here and isolate the problem because the code where this is happening is fairly complex and anytime I would try to simplify it to write a simple repro it would infuriatingly start working again...but I finally got it, and here it is:

```c#
void Main()
{	
        // Works fine
	SomeClass.Method1().Dump(); 

        // RuntimeBinderException: Cannot convert type 'System.Enum' to 'decimal'
	SomeClass.Method2().Dump();
}

public static class SomeClass {
	public static bool Method1()
	{
		return (decimal)(dynamic)Test.A == (decimal)(dynamic)Test.A;
	}
	
	public static bool Method2()
	{
		return CompareDecimalValues(Test.A, Test.A);
	}

	private enum Test
	{
		A = -100,
	}
}

public static bool CompareDecimalValues<T>(T value1, T value2) where T : unmanaged, Enum
{
	return (decimal)(dynamic)value1 == (decimal)(dynamic)value2;
}
```

If `CompareDecimalValues()` is inside the same assembly then changing the enum to internal or public makes `Method2` work. If `CompareDecimalValues` is outside the assembly then the enum must be changed to public for it to work. 

Basically, it seems like the method doing the dynamic cast must have visibility of the enum type otherwise it fails.

**Comment by mikernet** [comment]
Somewhat related: `Convert.ToDecimal` works for all cases.

**Comment by mikernet** [comment]
So it looks like this issue generalizes to any methods or operators (including the cast operators which is what caused the error above) on an object type that isn't visible to the executing context where dynamic is called. I guess that kind of makes sense.
